### PR TITLE
feat: Add SSM associations, patch baselines, patches

### PR DESCRIPTION
**Pull request by irmatov**

### CODE REVIEW

File: resources/provider/migrations/postgres/28_v0.10.20.down.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+DROP TABLE IF EXISTS aws_ssm_association_targets;
+DROP TABLE IF EXISTS aws_ssm_associations;
+
+-- Resource: ssm.instances
+DROP TABLE IF EXISTS aws_ssm_instance_patches;
+
+-- Resource: ssm.patch_baselines
+DROP TABLE IF EXISTS aws_ssm_patch_baseline_approval_rules;
+DROP TABLE IF EXISTS aws_ssm_patch_baseline_sources;
+DROP TABLE IF EXISTS aws_ssm_patch_baselines;
```
Comment (zagronitay) [inline resources/provider/migrations/postgres/28_v0.10.20.down.sql:12]: should be moved to 17

File: resources/provider/migrations/postgres/28_v0.10.20.up.sql [hunk 2]
```diff
@@ -0,0 +1,125 @@
+-- Resource: ssm.associations
```
Comment (zagronitay) [inline resources/provider/migrations/postgres/28_v0.10.20.up.sql:1]: should be moved to 17

File: resources/provider/migrations/timescale/28_v0.10.20.down.sql [hunk 3]
```diff
@@ -0,0 +1,12 @@
+-- Resource: ssm.associations
```
Comment (zagronitay) [inline resources/provider/migrations/timescale/28_v0.10.20.down.sql:1]: should be moved to 17

File: resources/provider/migrations/timescale/28_v0.10.20.up.sql [hunk 4]
```diff
@@ -0,0 +1,139 @@
+-- Resource: ssm.associations
```
Comment (zagronitay) [inline resources/provider/migrations/timescale/28_v0.10.20.up.sql:1]: should be moved to 17

File: resources/services/ssm/patch_baselines.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+						Type:        schema.TypeString,
+					},
+					{
+						Name:        "name",
+						Description: "The name specified to identify the patch source.  This member is required.",
+						Type:        schema.TypeString,
+					},
+					{
+						Name:        "products",
+						Description: "The specific operating system versions a patch repository applies to, such as \"Ubuntu16.04\", \"AmazonLinux2016.09\", \"RedhatEnterpriseLinux7.2\" or \"Suse12.7\". For lists of supported product values, see PatchFilter.  This member is required.",
```
Comment (zagronitay) [inline resources/services/ssm/patch_baselines.go:168]: remove the "this member is required"

File: resources/services/ssm/patch_baselines.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+						Resolver:    schema.ParentIdResolver,
+					},
+					{
+						Name:        "configuration",
+						Description: "The value of the yum repo configuration",
+						Type:        schema.TypeString,
+					},
+					{
+						Name:        "name",
+						Description: "The name specified to identify the patch source.  This member is required.",
```
Comment (zagronitay) [inline resources/services/ssm/patch_baselines.go:163]: remove the "this member is required"

File: resources/services/ssm/patch_baselines.go [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+func PatchBaselines() *schema.Table {
+	return &schema.Table{
+		Name:         "aws_ssm_patch_baselines",
+		Resolver:     fetchSsmPatchBaselines,
+		Multiplex:    client.ServiceAccountRegionMultiplexer("ssm"),
+		IgnoreError:  client.IgnoreAccessDeniedServiceDisabled,
+		DeleteFilter: client.DeleteAccountRegionFilter,
+		Options:      schema.TableCreationOptions{PrimaryKeys: []string{"account_id", "region", "baseline_id"}},
+		Columns: []schema.Column{
+			{
```
Comment (zagronitay) [inline resources/services/ssm/patch_baselines.go:25]: add arn 
https://docs.aws.amazon.com/service-authorization/latest/reference/list_awssystemsmanager.html

File: resources/services/ssm/patch_baselines.go [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+//go:generate cq-gen -config=patch_baselines.hcl -domain=ssm -resource=patch_baselines
+func PatchBaselines() *schema.Table {
+	return &schema.Table{
+		Name:         "aws_ssm_patch_baselines",
+		Resolver:     fetchSsmPatchBaselines,
+		Multiplex:    client.ServiceAccountRegionMultiplexer("ssm"),
+		IgnoreError:  client.IgnoreAccessDeniedServiceDisabled,
+		DeleteFilter: client.DeleteAccountRegionFilter,
+		Options:      schema.TableCreationOptions{PrimaryKeys: []string{"account_id", "region", "baseline_id"}},
```
Comment (zagronitay) [inline resources/services/ssm/patch_baselines.go:23]: arn should be the pk

File: resources/services/ssm/patch_baselines.go [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+	rule := resource.Item.(types.PatchRule)
+	if rule.PatchFilterGroup == nil {
+		return nil
+	}
+	data, err := json.Marshal(rule.PatchFilterGroup.PatchFilters)
+	if err != nil {
+		return diag.WrapError(err)
+	}
+	return diag.WrapError(resource.Set(c.Name, data))
+}
```
Comment (zagronitay) [inline resources/services/ssm/patch_baselines.go:249]: resource tags via
https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_ListTagsForResource.html

File: resources/services/ssm/associations.go [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+func fetchSsmAssociationTargetLocations(ctx context.Context, meta schema.ClientMeta, parent *schema.Resource, res chan<- interface{}) error {
+	a := parent.Item.(types.AssociationDescription)
+	res <- a.TargetLocations
+	return nil
+}
+func fetchSsmAssociationTargets(ctx context.Context, meta schema.ClientMeta, parent *schema.Resource, res chan<- interface{}) error {
+	a := parent.Item.(types.AssociationDescription)
+	res <- a.Targets
+	return nil
+}
```
Comment (zagronitay) [inline resources/services/ssm/associations.go:319]: resource tags via
https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_ListTagsForResource.html

File: resources/services/ssm/associations.go [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+	return &schema.Table{
+		Name:         "aws_ssm_associations",
+		Description:  "Describes the parameters for a document.",
+		Resolver:     fetchSsmAssociations,
+		Multiplex:    client.ServiceAccountRegionMultiplexer("ssm"),
+		IgnoreError:  client.IgnoreAccessDeniedServiceDisabled,
+		DeleteFilter: client.DeleteAccountRegionFilter,
+		Options:      schema.TableCreationOptions{PrimaryKeys: []string{"account_id", "region", "id"}},
+		Columns: []schema.Column{
+			{
```
Comment (zagronitay) [inline resources/services/ssm/associations.go:25]: add arn
https://docs.aws.amazon.com/service-authorization/latest/reference/list_awssystemsmanager.html

File: resources/services/ssm/associations.go [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+//go:generate cq-gen -config=associations.hcl -domain=ssm -resource=associations
+func Associations() *schema.Table {
+	return &schema.Table{
+		Name:         "aws_ssm_associations",
+		Description:  "Describes the parameters for a document.",
+		Resolver:     fetchSsmAssociations,
+		Multiplex:    client.ServiceAccountRegionMultiplexer("ssm"),
+		IgnoreError:  client.IgnoreAccessDeniedServiceDisabled,
+		DeleteFilter: client.DeleteAccountRegionFilter,
+		Options:      schema.TableCreationOptions{PrimaryKeys: []string{"account_id", "region", "id"}},
```
Comment (zagronitay) [inline resources/services/ssm/associations.go:23]: arn should be the pk

File: resources/services/ssm/associations.go [hunk 13]
```diff
@@ -0,0 +1,319 @@
...
+					},
+					{
+						Name:        "values",
+						Description: "User-defined criteria that maps to Key",
+						Type:        schema.TypeStringArray,
+					},
+				},
+			},
+		},
+	}
```
Comment (zagronitay) [inline resources/services/ssm/associations.go:271]: not sure but maybe can we change into JSON field

**Review by zagronitay — CHANGES_REQUESTED** [review]

File: resources/provider/migrations/postgres/28_v0.10.20.down.sql (hunk 1 above)
Comment (irmatov) [reply resources/provider/migrations/postgres/28_v0.10.20.down.sql:12]: Updated.

File: resources/provider/migrations/postgres/28_v0.10.20.up.sql (hunk 2 above)
Comment (irmatov) [reply resources/provider/migrations/postgres/28_v0.10.20.up.sql:1]: updated

File: resources/provider/migrations/timescale/28_v0.10.20.down.sql (hunk 3 above)
Comment (irmatov) [reply resources/provider/migrations/timescale/28_v0.10.20.down.sql:1]: updated

File: resources/provider/migrations/timescale/28_v0.10.20.up.sql (hunk 4 above)
Comment (irmatov) [reply resources/provider/migrations/timescale/28_v0.10.20.up.sql:1]: updated

File: resources/services/ssm/patch_baselines.go (hunk 5 above)
Comment (irmatov) [reply resources/services/ssm/patch_baselines.go:168]: done.

File: resources/services/ssm/patch_baselines.go (hunk 6 above)
Comment (irmatov) [reply resources/services/ssm/patch_baselines.go:163]: done

File: resources/services/ssm/associations.go (hunk 10 above)
Comment (irmatov) [reply resources/services/ssm/associations.go:319]: associations don't have tags

File: resources/services/ssm/patch_baselines.go (hunk 7 above)
Comment (irmatov) [reply resources/services/ssm/patch_baselines.go:25]: Done

File: resources/services/ssm/patch_baselines.go (hunk 8 above)
Comment (irmatov) [reply resources/services/ssm/patch_baselines.go:23]: Done

File: resources/services/ssm/patch_baselines.go (hunk 9 above)
Comment (irmatov) [reply resources/services/ssm/patch_baselines.go:249]: Added tags

File: resources/services/ssm/associations.go (hunk 11 above)
Comment (irmatov) [reply resources/services/ssm/associations.go:25]: Added

File: resources/services/ssm/associations.go (hunk 12 above)
Comment (irmatov) [reply resources/services/ssm/associations.go:23]: Done

File: resources/services/ssm/associations.go (hunk 13 above)
Comment (irmatov) [reply resources/services/ssm/associations.go:271]: I agree. Collapsed into a JSON field.

**Review by zagronitay — APPROVED** [review]

File: resources/services/ssm/associations.go [hunk 14]
```diff
@@ -0,0 +1,319 @@
...
+				},
+				func(o *ssm.Options) { o.Region = cl.Region },
+			)
+			if err != nil {
+				return diag.WrapError(err)
+			}
+			if desc.AssociationDescription != nil {
+				res <- *desc.AssociationDescription
+			}
+		}
```
Comment (bbernays) [inline resources/services/ssm/associations.go:292]: Not sure how many Associations accounts typically have. Do we want to put the DescribeAssociation call in a go routine? We have done it for a couple of instances like here:
https://github.com/cloudquery/cq-provider-aws/pull/701/files

File: resources/services/ssm/patch_baselines.go [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+					o.Region = cl.Region
+				},
+			)
+			if err != nil {
+				return diag.WrapError(err)
+			}
+			if b != nil {
+				res <- *b
+			}
+		}
```
Comment (bbernays) [inline resources/services/ssm/patch_baselines.go:224]: Same issue of performance and use of go routines

File: resources/services/ssm/associations.go (hunk 14 above)
Comment (irmatov) [reply resources/services/ssm/associations.go]: Made the calls concurrent as suggested.

File: resources/services/ssm/patch_baselines.go (hunk 15 above)
Comment (irmatov) [reply resources/services/ssm/patch_baselines.go]: Adjusted.

**Review by yevgenypats — CHANGES_REQUESTED** [review]
Looks good. Please move this to monorepo - https://github.com/cloudquery/cloudquery

**Pull request closed without merging**